### ISSUE TITLE
Weird intermittent hangs in tower after updating to 0.69

### ISSUE THREAD

**Issue by praveenperera** [kube-rs/kube-rs#829]
### Current and expected behavior

So this isn't going to be the best error report because the problem seems to be intermittent. But I'm going to make it anyway because maybe someone else might be having the same problem. 

I was on kube 0.64, everything was working fine, but when I upgraded to 0.69 I noticed that sometimes my k8s resources weren't being created. I enabled all debug logs are it seems to get stuck on. `tower::buffer::worker service.ready=true processing request`. I didn't have this issue locally. And its intermittent. 

Since then I have downgraded back to 0.64 and everything seems to be working fine.

Sorry that I couldn't be more helpful here. Maybe I should make this in the tower repo. Maybe this is only present after the tower-http 0.2 upgrade. When I have more time I will try a few different things to see if I can narrow it down.

Was curious if anyone else using kube-rs has seen a similar issue.

### Environment

* gke cluster (1.20.12-gke.1500)
* debian:bullseye-slim

### Affected crates

kube-client

### Would you like to work on fixing this bug?

no

**Comment by teozkr** [comment]
Hm, odd.

1. Can you confirm that the bug is new in 0.69, and does not happen with 0.68?
2. Would you mind running `cargo update` on the latest build of your app that does work fine for you (whether that ends up using kube-rs 0.64, 0.68, or something in between), and confirm that that still works?
3. Would you mind posting the `diff` between the `Cargo.lock`s of the newest working and oldest broken versions?

**Comment by clux** [comment]
I can reproduce it in `controller-rs` repo in main now: `tilt up` against `k3d`

after about a minute the controller stops handling messages and it prints a bunch of tower and hyper messages:

```
foo-controll… │ {"timestamp":"2022-02-16T16:54:52.700113Z","level":"TRACE","fields":{"message":"worker polling for next message"},"target":"tower::buffer::worker"}



foo-controll… │ {"timestamp":"2022-02-16T16:54:55.384825Z","level":"TRACE","fields":{"message":"decode; state=Chunked(Size, 0)"},"target":"hyper::proto::h1::decode"}
foo-controll… │ {"timestamp":"2022-02-16T16:54:55.384847Z","level":"TRACE","fields":{"message":"Read chunk hex size"},"target":"hyper::proto::h1::decode"}
foo-controll… │ {"timestamp":"2022-02-16T16:54:55.384875Z","level":"TRACE","fields":{"message":"received 905 bytes"},"target":"hyper::proto::h1::io"}
foo-controll… │ {"timestamp":"2022-02-16T16:54:55.384883Z","level":"TRACE","fields":{"message":"Read chunk hex size"},"target":"hyper::proto::h1::decode"}
foo-controll… │ {"timestamp":"2022-02-16T16:54:55.384888Z","level":"TRACE","fields":{"message":"Read chunk hex size"},"target":"hyper::proto::h1::decode"}
foo-controll… │ {"timestamp":"2022-02-16T16:54:55.384893Z","level":"TRACE","fields":{"message":"Read chunk hex size"},"target":"hyper::proto::h1::decode"}
foo-controll… │ {"timestamp":"2022-02-16T16:54:55.384897Z","level":"TRACE","fields":{"message":"Chunk size is 898"},"target":"hyper::proto::h1::decode"}
foo-controll… │ {"timestamp":"2022-02-16T16:54:55.384904Z","level":"DEBUG","fields":{"message":"incoming chunked header: 0x382 (898 bytes)"},"target":"hyper::proto::h1::decode"}
foo-controll… │ {"timestamp":"2022-02-16T16:54:55.384910Z","level":"TRACE","fields":{"message":"Chunked read, remaining=898"},"target":"hyper::proto::h1::decode"}
foo-controll… │ {"timestamp":"2022-02-16T16:54:55.384925Z","level":"TRACE","fields":{"message":"flushed({role=client}): State { reading: Body(Chunked(BodyCr, 0)), writing: KeepAlive, keep_alive: Busy }"},"target":"hyper::proto::h1::conn"}
foo-controll… │ {"timestamp":"2022-02-16T16:54:55.384940Z","level":"TRACE","fields":{"message":"flushed({role=client}): State { reading: Body(Chunked(BodyCr, 0)), writing: KeepAlive, keep_alive: Busy }"},"target":"hyper::proto::h1::conn"}
foo-controll… │ {"timestamp":"2022-02-16T16:54:55.384955Z","level":"TRACE","fields":{"message":"decode; state=Chunked(BodyCr, 0)"},"target":"hyper::proto::h1::decode"}
foo-controll… │ {"timestamp":"2022-02-16T16:54:55.384962Z","level":"TRACE","fields":{"message":"Read chunk hex size"},"target":"hyper::proto::h1::decode"}
foo-controll… │ {"timestamp":"2022-02-16T16:54:55.384974Z","level":"TRACE","fields":{"message":"flushed({role=client}): State { reading: Body(Chunked(Size, 0)), writing: KeepAlive, keep_alive: Busy }"},"target":"hyper::proto::h1::conn"}
foo-controll… │ {"timestamp":"2022-02-16T16:54:56.112946Z","level":"TRACE","fields":{"message":"decode; state=Chunked(Size, 0)"},"target":"hyper::proto::h1::decode"}
foo-controll… │ {"timestamp":"2022-02-16T16:54:56.112967Z","level":"TRACE","fields":{"message":"Read chunk hex size"},"target":"hyper::proto::h1::decode"}
foo-controll… │ {"timestamp":"2022-02-16T16:54:56.112992Z","level":"TRACE","fields":{"message":"received 903 bytes"},"target":"hyper::proto::h1::io"}
foo-controll… │ {"timestamp":"2022-02-16T16:54:56.112999Z","level":"TRACE","fields":{"message":"Read chunk hex size"},"target":"hyper::proto::h1::decode"}
foo-controll… │ {"timestamp":"2022-02-16T16:54:56.113003Z","level":"TRACE","fields":{"message":"Read chunk hex size"},"target":"hyper::proto::h1::decode"}
foo-controll… │ {"timestamp":"2022-02-16T16:54:56.113006Z","level":"TRACE","fields":{"message":"Read chunk hex size"},"target":"hyper::proto::h1::decode"}
foo-controll… │ {"timestamp":"2022-02-16T16:54:56.113009Z","level":"TRACE","fields":{"message":"Chunk size is 896"},"target":"hyper::proto::h1::decode"}
foo-controll… │ {"timestamp":"2022-02-16T16:54:56.113014Z","level":"DEBUG","fields":{"message":"incoming chunked header: 0x380 (896 bytes)"},"target":"hyper::proto::h1::decode"}
foo-controll… │ {"timestamp":"2022-02-16T16:54:56.113018Z","level":"TRACE","fields":{"message":"Chunked read, remaining=896"},"target":"hyper::proto::h1::decode"}
foo-controll… │ {"timestamp":"2022-02-16T16:54:56.113030Z","level":"TRACE","fields":{"message":"flushed({role=client}): State { reading: Body(Chunked(BodyCr, 0)), writing: KeepAlive, keep_alive: Busy }"},"target":"hyper::proto::h1::conn"}
foo-controll… │ {"timestamp":"2022-02-16T16:54:56.113066Z","level":"TRACE","fields":{"message":"decode; state=Chunked(BodyCr, 0)"},"target":"hyper::proto::h1::decode"}
foo-controll… │ {"timestamp":"2022-02-16T16:54:56.113074Z","level":"TRACE","fields":{"message":"Read chunk hex size"},"target":"hyper::proto::h1::decode"}
foo-controll… │ {"timestamp":"2022-02-16T16:54:56.113084Z","level":"TRACE","fields":{"message":"flushed({role=client}): State { reading: Body(Chunked(Size, 0)), writing: KeepAlive, keep_alive: Busy }"},"target":"hyper::proto::h1::conn"}
foo-controll… │ {"timestamp":"2022-02-16T16:54:56.113109Z","level":"TRACE","fields":{"message":"Non-error received, resetting backoff"},"target":"kube_runtime::utils::stream_backoff"}
foo-controll… │ {"timestamp":"2022-02-16T16:55:03.040278Z","level":"TRACE","fields":{"message":"decode; state=Chunked(Size, 0)"},"target":"hyper::proto::h1::decode"}
foo-controll… │ {"timestamp":"2022-02-16T16:55:03.040343Z","level":"TRACE","fields":{"message":"Read chunk hex size"},"target":"hyper::proto::h1::decode"}
foo-controll… │ {"timestamp":"2022-02-16T16:55:03.040448Z","level":"TRACE","fields":{"message":"received 118 bytes"},"target":"hyper::proto::h1::io"}
foo-controll… │ {"timestamp":"2022-02-16T16:55:03.040473Z","level":"TRACE","fields":{"message":"Read chunk hex size"},"target":"hyper::proto::h1::decode"}
foo-controll… │ {"timestamp":"2022-02-16T16:55:03.040487Z","level":"TRACE","fields":{"message":"Read chunk hex size"},"target":"hyper::proto::h1::decode"}
foo-controll… │ {"timestamp":"2022-02-16T16:55:03.040502Z","level":"TRACE","fields":{"message":"Chunk size is 112"},"target":"hyper::proto::h1::decode"}
foo-controll… │ {"timestamp":"2022-02-16T16:55:03.040516Z","level":"DEBUG","fields":{"message":"incoming chunked header: 0x70 (112 bytes)"},"target":"hyper::proto::h1::decode"}
foo-controll… │ {"timestamp":"2022-02-16T16:55:03.040542Z","level":"TRACE","fields":{"message":"Chunked read, remaining=112"},"target":"hyper::proto::h1::decode"}
foo-controll… │ {"timestamp":"2022-02-16T16:55:03.040581Z","level":"TRACE","fields":{"message":"flushed({role=client}): State { reading: Body(Chunked(BodyCr, 0)), writing: KeepAlive, keep_alive: Busy }"},"target":"hyper::proto::h1::conn"}
foo-controll… │ {"timestamp":"2022-02-16T16:55:03.040765Z","level":"TRACE","fields":{"message":"decode; state=Chunked(BodyCr, 0)"},"target":"hyper::proto::h1::decode"}
foo-controll… │ {"timestamp":"2022-02-16T16:55:03.040804Z","level":"TRACE","fields":{"message":"Read chunk hex size"},"target":"hyper::proto::h1::decode"}
foo-controll… │ {"timestamp":"2022-02-16T16:55:03.040847Z","level":"TRACE","fields":{"message":"flushed({role=client}): State { reading: Body(Chunked(Size, 0)), writing: KeepAlive, keep_alive: Busy }"},"target":"hyper::proto::h1::conn"}
```

**Comment by teozkr** [comment]
That would make sense with the minute delay too, actually...

Does the reloading affect our public API? Do we want to release a 0.69.1 with it reverted for now?

**Comment by clux** [comment]
It's an entirely private change - let's spend a little bit of time to see if we can spot the cause in the PR before releasing a fix. It's simple to repro, and the PR is small.

**Comment by praveenperera** [comment]
If I had to guess I would think its blocking here:

```
                if let Some(token) = token_file.read().await.cached_token() {
                     bearer_header(token)
                 } else {
                     bearer_header(token_file.write().await.token())
                 }
```

**Comment by clux** [comment]
think i got it. swapping the rwlock for a mutex and taking the lock outside that branch start seems to stop it from happening

**Comment by praveenperera** [comment]
Thanks, you guys are awesome from issue report to fix in within an hour. Appreciate it. Using kube-rs has been awesome 😊

**Comment by teozkr** [comment]
0.69.1 is now released, and contains the fix!

**Comment by praveenperera** [comment]
0.69.1 confirmed working in cluster. Thanks!

**Comment by kazk** [comment]
Sorry for this :( Thanks for getting the fix released quickly @clux @teozkr.

I thought I had tested it, but maybe that was only with the initial version that used `Mutex`. We can also remove `cached_token()` with `Mutex`.

---

What can we do to prevent this in the future? Always update and test `controller-rs` before release?

**Comment by clux** [comment]
Yeah, it's not really practical to run all tests in-cluster (and even if we had run tests in-cluster before release, we would not have caught this because the unit test does not exercise the layer in parallel). It's also not to practical to always run `controller-rs` as it's our convenience + example rather than a CI enforced thing (it's a nice thing for us run for us when we do change in-cluster things regardless though) - and it's hard + overkill to make it a CI enforced thing.

E2E tests sit at the top of our testing pyramid to explicitly catch issues related to in-cluster changes. We have just one test in that category (awkwardly named `dapp`), and it's there just to make sure in-cluster auth works. I think we should work on the e2e suite so that we can test that the in-cluster auth works across reload points.

I don't really want to move too many tests in-cluster unless there are specific cases we cannot catch with lower level tests. E2E tests, while powerful, are also the most cumbersome to write, and debug when they break down as they suffers the most from CI interaction effects and slow execution times.

Stronger `Controller` testing is a separate issue I think we should also work on - albeit not related to this - as we currently are lacking [coverage of this module](https://app.codecov.io/gh/kube-rs/kube-rs/) even in the land of unit/integration tests. We should focus on that separately.
